src/common/procedure/src/local/runner.rs

```diff
@@ -217,16 +237,18 @@ impl Runner {
                 );
 
                 if e.is_retry_later() {
+                    self.meta.set_state(ProcedureState::retry(Arc::new(e)));
                     return ExecResult::RetryLater;
                 }
 
-                self.meta.set_state(ProcedureState::failed(Arc::new(e)));
-
                 // Write rollback key so we can skip this procedure while recovering procedures.
-                if self.rollback_procedure().await.is_err() {
+                if let Err(e) = self.rollback_procedure().await {
```

Comment:
issue: I realize that if we set state after rollback, then we will execute the procedure again if we return `RetryLater` here.

Maybe we could add a rolling_back flag in the runner to avoid `execute()` again.